### PR TITLE
🐛 [source-linkedin-ads] fix re-adding sponsored content stream

### PR DESCRIPTION
**Pull request by FVidalCarneiro**
## What
This pull request recovers the removed Ad Sponsored Content stream, and therefore solve the issue described [here](https://github.com/airbytehq/airbyte/issues/35635). To summarize the issue, this stream was removed without explanation and ingested data from an important and still usable Linkedin endpoint (doc [here](https://docs.microsoft.com/en-us/linkedin/marketing/integrations/ads/advertising-targeting/create-and-manage-video?tabs=http#finders)). 

Fixes (https://github.com/airbytehq/airbyte/issues/35635)

## How
To add this stream, the removed code was recovered and adapted to be compatible with the current version of the code.

## Recommended reading order
1. `streams.py`
2. `source.py`
3. `test_source.py`
4. `ad_direct_sponsored_contents.json`

## 🚨 User Impact 🚨
No breaking changes, other streams were not effected, only one was added.

This change should therefore be a patch (if seen as a bug fix) or a minor change, if seen as a new feature added.

## Pre-merge Actions

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206685813413289

### CODE REVIEW

**Comment by FVidalCarneiro** [comment]
Hi @artem1205 , thank you for your time to take a look at our PR. The documentation is not at all clear regarding this endpoint, I have updated the documentation links with a more useful link, you can verify the existence of this endpoint as well through this link.

The endpoint does exist and is working today. Please find below a screenshot of an example API call to the given endpoint using postman:

proof_endpoint_exists

I have likewise tested the code with a local run on our data and the data is collected as expected, we collect multiple elements and each element has the a schema of the following example record:

```
{
        "duration": 15000,
        "aspectRatioWidth": 9.0,
        "owner": "urn:li:organization:111111",
        "thumbnail": "https://link.com/path",
        "downloadUrlExpiresAt": 1711627200000,
        "downloadUrl": "https://link-2.com/path",
        "aspectRatioHeight": 16.0,
        "id": "urn:li:video:5555555",
        "mediaLibraryMetadata": {
            "associatedAccount": "urn:li:sponsoredAccount:222222",
            "mediaLibraryStatus": "ACTIVE",
            "assetName": "asset-name-example.mp4"
}
```

File: airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/schemas/ad_direct_sponsored_contents.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
+    "type": ["null", "object"],
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/schemas/ad_direct_sponsored_contents.json:2]: missing additional properties = true and json header. Please take a look in other schemas to fix that.

File: airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/schemas/ad_direct_sponsored_contents.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      "created": {
+        "type": ["null", "string"],
+        "format": "date-time"
+      },
+      "lastModified": {
+        "type": ["null", "string"],
+        "format": "date-time"
+      }
+    }
+  }
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/schemas/ad_direct_sponsored_contents.json:31]: EOF. Probably you can run `airbyte-ci format fix all` to help you.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Some comments.

**Comment by FVidalCarneiro** [comment]
@artem1205 , did you have time to verify that the endpoint exists by taking a look at my comment above ? @marcosmarxm , are there any other blockers for this PR to be merged ?

**Comment by FVidalCarneiro** [comment]
Hello @natikgadzhi , as promised, here is a description of why we need this endpoint. 

Essentially, Linkedin adverts are obtained through the endpoint Linkedin Creatives. Creatives, reference some sort of content, like videos through a field called `reference` (more information on this on the creatives endpoint [documentation](https://learn.microsoft.com/en-us/linkedin/marketing/integrations/ads/account-structure/create-and-manage-creatives?view=li-lms-2024-03&tabs=http%2Chttp-update-a-creative)). This is the field we use to make a join with the adSponsoredContent object, field `content_reference`. We need to make this join in order to obtain the name of the video that is associated to the Linkedin Creative we are analyzing. The adSponsoredContent is currently the only endpoint we could find that can give us the information necessary to make this join / connection. 

The documentation of this adSponsoredContent object is available [here](https://learn.microsoft.com/en-us/linkedin/marketing/integrations/ads/account-structure/create-and-manage-creatives?view=li-lms-2024-03&tabs=http%2Chttp-update-a-creative).

**Comment by natikgadzhi** [comment]
> Essentially, Linkedin adverts are obtained through the endpoint Linkedin Creatives. Creatives, reference some sort of content, like videos through a field called reference (more information on this on the creatives endpoint [documentation](https://learn.microsoft.com/en-us/linkedin/marketing/integrations/ads/account-structure/create-and-manage-creatives?view=li-lms-2024-03&tabs=http%2Chttp-update-a-creative)). This is the field we use to make a join with the adSponsoredContent object, field content_reference. We need to make this join in order to obtain the name of the video that is associated to the Linkedin Creative we are analyzing. The adSponsoredContent is currently the only endpoint we could find that can give us the information necessary to make this join / connection.

> The documentation of this adSponsoredContent object is available [here](https://learn.microsoft.com/en-us/linkedin/marketing/integrations/ads/account-structure/create-and-manage-creatives?view=li-lms-2024-03&tabs=http%2Chttp-update-a-creative).

Can you add this to the readme + docs for this connector? I think this will be useful for others, plus we're migrating the connector to low-code and I would love to not miss this. 

Additionally, @FVidalCarneiro when we will migrate this to low code, would you like to be a reviewer on our PR? /cc @lazebnyi

**Review by natikgadzhi — COMMENTED** [review]
This change seems to be additive, the code looks good to me. 

Approving, but I need a few things to be happy:
1. @FVidalCarneiro please add the notes you made to connector documentation. 
2. @lazebnyi who is working on linkin-ads low-code? I would like that engineer to perform live tests on this PR and attach screenshots of the report here. Let's do it this week. /cc @katmarkham @oustynova 
3. @alafanechere, please merge the CI on forks thing and trigger CI runs on this PR. When you do, please ping me with a link to this PR, I want to see it in action.

**Comment by FVidalCarneiro** [comment]
@natikgadzhi , as requested, I have added documentation regarding the ad sponsored contents stream. Let me know if you want us to add more details

**Comment by natikgadzhi** [comment]
@FVidalCarneiro, I've kicked off the CI run on this! Thank you for continuing to work on it. 

I see that formatters errored out on `source_linkedin_ads/source.py` - run `airbyte-ci format fix all` locally, or run `poetry run black` just for this file from the repo root.

**Comment by natikgadzhi** [comment]
Kicking off the CI, let's see what's going on here.

**Comment by FVidalCarneiro** [comment]
Hi @natikgadzhi , thanks, just pushed another commit to fix formatting errors (apologies for not having done it earlier). Tests seem to pass now !

**Comment by FVidalCarneiro** [comment]
@natikgadzhi , just solved the merge conflicts

**Comment by natikgadzhi** [comment]
I'll review the rest, but you'd need to `airbyte-ci format fix all` in your fork, our format-fix can't commit back to your fork ;)

**Comment by marcosmarxm** [comment]
Running CI @FVidalCarneiro let's wait for results.

**Comment by natikgadzhi** [comment]
```
process "python /airbyte/integration_code/main.py spec" did not complete successfully: exit code: 1
Stdout:
{"type": "LOG", "log": {"level": "FATAL", "message": "Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided.\nTraceback (most recent call last):\n  File \"/airbyte/integration_code/main.py\", line 8, in <module>\n    run()\n  File \"/airbyte/integration_code/source_linkedin_ads/run.py\", line 14, in run\n    launch(source, sys.argv[1:])\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 235, in launch\n    for message in source_entrypoint.run(parsed_args):\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 97, in run\n    source_spec: ConnectorSpecification = self.source.spec(self.logger)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/connector.py\", line 86, in spec\n    raise RuntimeError(\"Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided.\")\nRuntimeError: Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided."}}
{"type": "TRACE", "trace": {"type": "ERROR", "emitted_at": 1719414714121.64, "error": {"message": "Something went wrong in the connector. See the logs for more details.", "internal_message": "Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided.", "stack_trace": "Traceback (most recent call last):\n  File \"/airbyte/integration_code/main.py\", line 8, in <module>\n    run()\n  File \"/airbyte/integration_code/source_linkedin_ads/run.py\", line 14, in run\n    launch(source, sys.argv[1:])\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 235, in launch\n    for message in source_entrypoint.run(parsed_args):\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 97, in run\n    source_spec: ConnectorSpecification = self.source.spec(self.logger)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/connector.py\", line 86, in spec\n    raise RuntimeError(\"Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided.\")\nRuntimeError: Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided.\n", "failure_type": "system_error", "stream_descriptor": null}}}

```

I see you've added spec.json — probably change things in spec.yaml instead.

**Comment by marcosmarxm** [comment]
@FVidalCarneiro here you can access the [report](https://ff651a65f177ac9f1ab8d7e13ea82d67c905cef240c79ec3e84a68a-apidata.googleusercontent.com/download/storage/v1/b/airbyte-ci-reports-multi/o/airbyte-ci%2Fconnectors%2Ftest%2Fpull_request%2Ffvidalcarneiro_fix-re-adding-sponsored-content-stream%2F1719414285%2F17bca21d4fcd6a115ad5a34c617e401076ea4c46%2Fsource-linkedin-ads%2F2.1.7%2Foutput.html?jk=AS1gM8O-kCUsA64oyJIxfwW7CYDv8lFYjuTl2ZYQ1yRqjgrYS9bLqG6Z8of3-PbvvYeKFNTYz7_lgGaaZ2h2kJCQh6VW7ALGK02Z3vq-f5qW0wFr-H0dDCBSSP3jut9wIJIrK9Z_dUwZAPFvSXY7iw93OMqK8FHlzqOskyWcwP6oEyXkFNUEtCZaBSvX0mBLBb3661Azt7UupQG3lgVu7TL4cTUJECGhYJDAkROL29Cnxabrou_enCBPwwpR9MsC8s4mtj65SQVfMAQvrHWVL-LrXZTPzuN6eaLJbhFAos80CTYXyLsg_FeH-DQPE8nyCKc2UU4oGOwNil1nQy5jbsDbBwJ7sItFo4NLuKGdF3g7Y1cKH2xi1B5JwvbZs_bKOLjhsOibueRFXU_kxiZI5_uYjLyMVjgDm580oBT58uoVSy2maEzZzyiuVb1xUGrFDIMC9Os_a6_xa_vCkaP38shJJAUylumqtK5oKWT4JwNhH7IIHN00TZwyGtsNtba9BS6LUBLhAgowxZsfw1PP-zNz0__oTiRowQZjzqAw0RkY_2vl4cHtAgL_bvmEzjxbbaz_o5E83kV6_U2wM6KATlAqlkgHyKNooSKJPHf5l1IqROndmhZFos7lUuT0XIcWjRztpvcrD3uY0A-QJQrSEqrbZDso0x9duqYiyrtqYq_ZoKdNocKrM4JPBW_p_NQYVElHTvsr8_Y_iZKdHy1opYufz4E0H_-PiAbtRdbPkZTCtamSuSeEnu87xHK2YLvAYMyyFdphSr2CoNPZ2hPyZIMrpmGF4KdoJ0ZKc1dk0M8H8w8taP1JfRbcPCdYF9huT77eVu9Yo3NuQ2zRtt0OUgDjj8MZlJdRk9g6m4wVCA2WUzA77k3zG9eoCUzn7rLMIuxYXelb3AQCsQtnv828mIzBhn7CkEZOW6hJyvKrvMLcQ2sa5yyC3wATxlHhaNJdMiIM1Tj1WS3HsOxJ7yzxOGRGXLA4e-YsGeby8HdDe6FLp-DaO_cOGFMn1YqygZciMXmyp1TlruHpCe5CGMNnTYNPX_TT9mHaqOFgCJIBzEZCZhZ6KCBDjfViN40oL9Awep2J465sc3oTpOQQVzXIOUKJQdFQXIVO5IURGrytv5mv83GPMWuaDSyJqXIBQF4myhp6YFZldwPQQIPBLJ6_tSpY4HUJI_l4iZNi4cnoKnxLw-D5iGlyQgR8Rgbw2IjHu7iTfPDqwJFa9xHZzNAt8hh_bseucmM6IwZ6VCFMPSewJk8PFFxnxRJdXADigs1J2KLIKWFdwIeT52a0PWe3rNp_Hh7EKEHk30tJMrVrTzofhU-zKlJQycUGbgMxAETjtmZHKqAv2FUp9_mj1EQZNrxLMdTHTSg2mReclBRa1Zts8BX7H483LSYWAGM9ldVUwq9kzyfHtVDeGuYzOwGdC8yH5IaAHAbfnOvJ1qGZiKB1eQbxn0SAiZRrMykmDacwfcicD5k&isca=1)

**Comment by FVidalCarneiro** [comment]
> ```
> process "python /airbyte/integration_code/main.py spec" did not complete successfully: exit code: 1
> Stdout:
> {"type": "LOG", "log": {"level": "FATAL", "message": "Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided.\nTraceback (most recent call last):\n  File \"/airbyte/integration_code/main.py\", line 8, in <module>\n    run()\n  File \"/airbyte/integration_code/source_linkedin_ads/run.py\", line 14, in run\n    launch(source, sys.argv[1:])\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 235, in launch\n    for message in source_entrypoint.run(parsed_args):\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 97, in run\n    source_spec: ConnectorSpecification = self.source.spec(self.logger)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/connector.py\", line 86, in spec\n    raise RuntimeError(\"Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided.\")\nRuntimeError: Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided."}}
> {"type": "TRACE", "trace": {"type": "ERROR", "emitted_at": 1719414714121.64, "error": {"message": "Something went wrong in the connector. See the logs for more details.", "internal_message": "Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided.", "stack_trace": "Traceback (most recent call last):\n  File \"/airbyte/integration_code/main.py\", line 8, in <module>\n    run()\n  File \"/airbyte/integration_code/source_linkedin_ads/run.py\", line 14, in run\n    launch(source, sys.argv[1:])\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 235, in launch\n    for message in source_entrypoint.run(parsed_args):\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 97, in run\n    source_spec: ConnectorSpecification = self.source.spec(self.logger)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/connector.py\", line 86, in spec\n    raise RuntimeError(\"Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided.\")\nRuntimeError: Found multiple spec files in the package. Only one of spec.yaml or spec.json should be provided.\n", "failure_type": "system_error", "stream_descriptor": null}}}
> ```
> 
> I see you've added spec.json — probably change things in spec.yaml instead.

Thanks @natikgadzhi , my bad, removed the spec.json file keeping the yaml file. You can go ahead and relaunch the CI

**Comment by FVidalCarneiro** [comment]
> @FVidalCarneiro here you can access the [report](https://ff651a65f177ac9f1ab8d7e13ea82d67c905cef240c79ec3e84a68a-apidata.googleusercontent.com/download/storage/v1/b/airbyte-ci-reports-multi/o/airbyte-ci%2Fconnectors%2Ftest%2Fpull_request%2Ffvidalcarneiro_fix-re-adding-sponsored-content-stream%2F1719414285%2F17bca21d4fcd6a115ad5a34c617e401076ea4c46%2Fsource-linkedin-ads%2F2.1.7%2Foutput.html?jk=AS1gM8O-kCUsA64oyJIxfwW7CYDv8lFYjuTl2ZYQ1yRqjgrYS9bLqG6Z8of3-PbvvYeKFNTYz7_lgGaaZ2h2kJCQh6VW7ALGK02Z3vq-f5qW0wFr-H0dDCBSSP3jut9wIJIrK9Z_dUwZAPFvSXY7iw93OMqK8FHlzqOskyWcwP6oEyXkFNUEtCZaBSvX0mBLBb3661Azt7UupQG3lgVu7TL4cTUJECGhYJDAkROL29Cnxabrou_enCBPwwpR9MsC8s4mtj65SQVfMAQvrHWVL-LrXZTPzuN6eaLJbhFAos80CTYXyLsg_FeH-DQPE8nyCKc2UU4oGOwNil1nQy5jbsDbBwJ7sItFo4NLuKGdF3g7Y1cKH2xi1B5JwvbZs_bKOLjhsOibueRFXU_kxiZI5_uYjLyMVjgDm580oBT58uoVSy2maEzZzyiuVb1xUGrFDIMC9Os_a6_xa_vCkaP38shJJAUylumqtK5oKWT4JwNhH7IIHN00TZwyGtsNtba9BS6LUBLhAgowxZsfw1PP-zNz0__oTiRowQZjzqAw0RkY_2vl4cHtAgL_bvmEzjxbbaz_o5E83kV6_U2wM6KATlAqlkgHyKNooSKJPHf5l1IqROndmhZFos7lUuT0XIcWjRztpvcrD3uY0A-QJQrSEqrbZDso0x9duqYiyrtqYq_ZoKdNocKrM4JPBW_p_NQYVElHTvsr8_Y_iZKdHy1opYufz4E0H_-PiAbtRdbPkZTCtamSuSeEnu87xHK2YLvAYMyyFdphSr2CoNPZ2hPyZIMrpmGF4KdoJ0ZKc1dk0M8H8w8taP1JfRbcPCdYF9huT77eVu9Yo3NuQ2zRtt0OUgDjj8MZlJdRk9g6m4wVCA2WUzA77k3zG9eoCUzn7rLMIuxYXelb3AQCsQtnv828mIzBhn7CkEZOW6hJyvKrvMLcQ2sa5yyC3wATxlHhaNJdMiIM1Tj1WS3HsOxJ7yzxOGRGXLA4e-YsGeby8HdDe6FLp-DaO_cOGFMn1YqygZciMXmyp1TlruHpCe5CGMNnTYNPX_TT9mHaqOFgCJIBzEZCZhZ6KCBDjfViN40oL9Awep2J465sc3oTpOQQVzXIOUKJQdFQXIVO5IURGrytv5mv83GPMWuaDSyJqXIBQF4myhp6YFZldwPQQIPBLJ6_tSpY4HUJI_l4iZNi4cnoKnxLw-D5iGlyQgR8Rgbw2IjHu7iTfPDqwJFa9xHZzNAt8hh_bseucmM6IwZ6VCFMPSewJk8PFFxnxRJdXADigs1J2KLIKWFdwIeT52a0PWe3rNp_Hh7EKEHk30tJMrVrTzofhU-zKlJQycUGbgMxAETjtmZHKqAv2FUp9_mj1EQZNrxLMdTHTSg2mReclBRa1Zts8BX7H483LSYWAGM9ldVUwq9kzyfHtVDeGuYzOwGdC8yH5IaAHAbfnOvJ1qGZiKB1eQbxn0SAiZRrMykmDacwfcicD5k&isca=1)

@marcosmarxm , link expired 🤔

**Review by natikgadzhi — APPROVED** [review]
I like this, and I think it's getting close to ready. Once the CI is green, I'm ok. 

@lazebnyi want to give this a quick review, make sure I haven't missed any Python antipatterns?

**Comment by FVidalCarneiro** [comment]
@marcosmarxm , as described by @natikgadzhi, I believe the last test is on your end to solve (regarding the sandbox account and `acceptance-test-config`). Could you please confirm this ? Also, I see that this PR still has your request for changes that I think we responded to, can you confirm this ?

**Comment by marcosmarxm** [comment]
The error is:
> E               Failed: High strictness level error: posts streams are declared in the catalog but do not have expected records. Please add expected records to integration_tests/expected_records.jsonl or declare these streams in empty_streams.

We need to add the expected records this stream is reading from sandbox account.

**Comment by FVidalCarneiro** [comment]
> The error is:
> 
> > E               Failed: High strictness level error: posts streams are declared in the catalog but do not have expected records. Please add expected records to integration_tests/expected_records.jsonl or declare these streams in empty_streams.
> 
> We need to add the expected records this stream is reading from sandbox account.

@marcosmarxm , ok so do you want us to add some fictive records or will you add some from your sandbox account ?

**Comment by marcosmarxm** [comment]
@FVidalCarneiro I think now is up to us fix some tests requires access to Airbyte sandbox.

> {"type": "RECORD", "record": {"stream": "posts", "data": {"lifecycleState": "PUBLISHED", "lastModifiedAt": 1629888842072, "visibility": "PUBLIC", "publishedAt": 1629888842026, ..., "emitted_at": 1720101201826}}
{"type": "RECORD", "record": {"stream": "posts", "data": {"lifecycleState": "PUBLISHED", "lastModifiedAt": 1629665367207, "visibility": "PUBLIC", "publishedAt": 1629665367207, "author":...}, "emitted_at": 1720101202178}}
{"type": "LOG", "log": {"level": "ERROR", "message": "{\"code\":\"FORBIDDEN\",\"message\":\"Accessing the resource is forbidden. Please check your permissions for this resource\",\"status\":403}"}}

@natikgadzhi, I ran these changes locally and was able to retrieve two records using OAuth. However, it fails afterward. I'm not sure if we need to update the permissions in the OAuth scope or introduce error handling in the connector code itself. Can you assist me with this? I tried to use the Token and couldn't read any records with that credential.

**Comment by kev-datams** [comment]
@FVidalCarneiro, probably need for adapting this stream to the new low-code linkedin-ads ?

**Comment by kev-datams** [comment]
hello @marcosmarxm @natikgadzhi , please coud you reopen the PR ?

**Pull request closed without merging**